sql/2020/11_Security/csp_number_of_allowed_hosts.sql

```diff
@@ -0,0 +1,41 @@
+#standardSQL
+# CSP on home pages: number of unique headers, header length and number of allowed hosts in all directives
+CREATE TEMPORARY FUNCTION getHeader(headers STRING, headername STRING)
+  RETURNS STRING DETERMINISTIC
+  LANGUAGE js AS '''
+  const parsed_headers = JSON.parse(headers);
+  const matching_headers = parsed_headers.filter(h => h.name.toLowerCase() == headername.toLowerCase());
+  if (matching_headers.length > 0) {
+    return matching_headers[0].value;
+  }
+  return null;
+''';
+CREATE TEMP FUNCTION array_unique(arr ANY TYPE) AS (
+  (SELECT ARRAY_AGG(DISTINCT x) FROM UNNEST(arr) as x)
+);
+
+SELECT
+  client,
+  percentile,
+  COUNT(0) AS total,
+  COUNTIF(csp_header IS NOT NULL) AS num_csp_headers,
+  COUNTIF(csp_header IS NOT NULL) / COUNT(0) AS pct_csp_headers,
+  COUNT(DISTINCT csp_header) AS num_unique_csp_headers,
+  APPROX_QUANTILES(LENGTH(csp_header), 1000 IGNORE NULLS)[OFFSET(percentile * 10)] AS csp_header_length,
+  APPROX_QUANTILES(ARRAY_LENGTH(array_unique(REGEXP_EXTRACT_ALL(csp_header, r'(?i)\W(https*://[^\s;]+)[\s;]'))), 1000 IGNORE NULLS)[OFFSET(percentile * 10)] AS unique_allowed_hosts,
```

Comment:
Quite difficult to read this. Since you're using a function(`arrray_unique`) already should you pull more of this out into that function?